website/docs/guides/version-4-upgrade.html.md

```diff
@@ -0,0 +1,223 @@
+---
+subcategory: ""
+layout: "aws"
+page_title: "Terraform AWS Provider Version 4 Upgrade Guide"
+description: |-
+  Terraform AWS Provider Version 4 Upgrade Guide
+---
+
+# Terraform AWS Provider Version 4 Upgrade Guide
+
+Version 4.0.0 of the AWS provider for Terraform is a major release and includes some changes that you will need to consider when upgrading. This guide is intended to help with that process and focuses only on changes from version 3.X to version 4.0.0. See the [Version 3 Upgrade Guide](/docs/providers/aws/guides/version-3-upgrade.html) for information about upgrading from 1.X to version 3.0.0.
+
+Most of the changes outlined in this guide have been previously marked as deprecated in the Terraform plan/apply output throughout previous provider releases. These changes, such as deprecation notices, can always be found in the [Terraform AWS Provider CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md).
+
+~> **NOTE:** Version 4.0.0 of the AWS Provider will be the last major version to support [EC2-Classic resources](#ec2-classic-resource-and-data-source-support) as AWS plans to fully retire EC2-Classic Networking. See the [AWS News Blog](https://aws.amazon.com/blogs/aws/ec2-classic-is-retiring-heres-how-to-prepare/) for additional details.
+
+~> **NOTE:** Version 4.0.0 and 4.X versions of the AWS Provider will be the last versions compatible with Terraform 0.12-0.14.
```

Comment:
From some internal discussions:

##### Question

> We are preparing to release v4.0.0 of the AWS Provider (hopefully this Thursday) and in the Upgrade Guide (https://github.com/hashicorp/terraform-provider-aws/pull/22818) have this note
> Version 4.0.0 and 4.X versions of the AWS Provider will be the last versions compatible with Terraform 0.12-0.14.
> This to enable us to support Protocol v6 from our next major version (v5.0.0) onwards.
> Realistically we will have a (long) period of time where we use a muxed solution between Plugin SDK v2 and vNext, so does this PR https://github.com/hashicorp/terraform/pull/30379 mean that actually v1.1.5 will be the minimum Core version required for any future muxed provider?

##### Answer

> Terraform CLI 1.1.5 and later will be required for upgrading sdk/v2 to protocol version 6 and muxing with framework.
> Other provider combination scenarios are not affected:
> Downgrading framework to protocol version 5 and muxing with sdk/v2
> Muxing protocol version 5 servers
> Muxing protocol version 6 servers
> So rather than Terraform CLI 1.0.3 and later being required for full protocol version 6 compatibility, its 1.1.5 and later in that specific use case.
> The reason it wasn’t backported to 1.0.x is because Terraform’s 1.0 compatibility guarantees means that we will always just recommend the latest Terraform 1.x release, rather than say, Terraform 1.0.x, 1.1.x, etc.

Reply:
I'd recommend saying 0.12 through 0.15 (or prior to 1.0), if the goal is that the next major version will be protocol version 6 (in some fashion).

```suggestion
~> **NOTE:** Version 4.0.0 and 4.X versions of the AWS Provider will be the last versions compatible with Terraform 0.12-0.15.
```

Reply:
Yes, I was thinking the same thing.